File: Stack/BOJ2504.py
def _2504():
    Stack = []
    N = str(input())

    cur = 1
    res = 0

    for i in range(len(N)):
        
        if N[i] == ']' or N[i] ==')':
            if not Stack:
                print(0)
                return
            
        if N[i] == '(':
            Stack.append('(')
            cur = cur * 2
            if i + 1 < len(N) and N[i+1] == ')': # 길이체크, 닫히는거면 결과에 더해줘
                # print(cur)
                res += cur

        elif N[i] == '[':
            Stack.append('[')
            cur = cur * 3
            if i + 1 < len(N) and N[i+1] == ']':
                res += cur

        if Stack:
            if N[i] == ')':
                cur = cur // 2
                if Stack[-1] == '(':
                    Stack.pop()
            elif N[i] == ']':
                cur = cur // 3
                if Stack[-1] == '[':
                    Stack.pop()
        print('curr' , N[i], 'stack', Stack, 'curr', cur, 'result', res)


    if Stack:
        print(0)
        return 0

    print(res)
_2504()

File: Stack/BOJ2504.java
package Stack;
import java.util.*;

public class BOJ2504 {
    public static void main(String[] Args){
        Scanner sc = new Scanner(System.in);
        String input = sc.next();
        Stack<Character> st = new Stack<>();
        int curr = 1;
        int result = 0;
        for(int i = 0 ; i < input.length() ; i++){
            Character tmp = input.charAt(i);

            if (tmp == '('  ){
                st.push(tmp);
                curr *= 2;
            }else if(tmp == '[') {
                st.push(tmp);
                curr *= 3;
            }else if(tmp == ')' ){
                if(st.isEmpty() || st.peek() != '('){
                    result = 0;
                    break;
                }
                if(input.charAt(i-1) == '('){
                    result += curr;
                }
                st.pop();
                curr /= 2;
            }else if( tmp == ']' ){
                if(st.isEmpty() || st.peek() != '['){
                    result = 0;
                    break;
                }
                if(input.charAt(i-1) == '['){
                    result += curr;
                }
                st.pop();
                curr /= 3;
            }
        }
        System.out.println(!st.isEmpty()? 0:result);
    }
}



// package Stack;

// import java.util.*;

// public class BOJ2504 {
//     public static void main(String[] Args){
//         Scanner sc = new Scanner(System.in);
//         String input = sc.nextLine();
//         Stack<Character> st = new Stack<Character>();

//         int curr = 1;
//         int result = 0;

//         for(int i = 0 ; i < input.length() ; i++){
//             Character tmp = input.charAt(i);
//             switch(tmp) {

//             case '(':
//                 st.push('(');
//                 curr *= 2;
//                 break;
//             case '[':
//                 st.push('[');
//                 curr *= 3;
//                 break;
//             case ')':
//                 if(st.isEmpty() || st.peek() != '('){
//                     result = 0;
//                     break;
//                 }
//                 if(input.charAt(i-1) == '('){
//                     result += curr;
//                 }
//                 st.pop();
//                 curr /= 2;
//                 break;
//             case ']':
//                 if(st.isEmpty() || st.peek() != '['){
//                     result = 0;
//                     break;
//                 }
//                 if(input.charAt(i-1) == '['){
//                     result += curr;
//                 }
//                 st.pop();
//                 curr /= 3;
//                 break;
//             }
//         }
// 		System.out.println(!St.isEmpty()? 0:result);
//     }
// }
